sql/2020/03_Markup/pages_almanac_by_device.sql

```diff
@@ -0,0 +1,111 @@
+#standardSQL
+# pages almanac metrics grouped by device
+
+# real run estimated at $4.08 and took 48 seconds
+
+# to speed things up there is only one js function per custom metric property. It returns a STRUCT with all the data needed
+# current test gathers 3 bits of incormation from the custom petric properties
+# I tried to do a single js function processing the whole payload but it was very slow (50 sec) because of parsing the full payload in js
+# this uses JSON_EXTRACT_SCALAR to first get the custom metrics json string, and only passes those into the js functions
+# Estimate about twice the speed of the original code. But should scale up far better as the custom metrics are only parsed once.
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+# returns all the data we need from _almanac
+CREATE TEMPORARY FUNCTION get_almanac_info(almanac_string STRING)
+RETURNS STRUCT<
+  scripts_total INT64,
+  none_jsonld_scripts_total INT64,
+  src_scripts_total INT64,
+  inline_scripts_total INT64,
+  good_heading_sequence BOOL,
+  contains_videos_with_autoplay BOOL,
+  contains_videos_without_autoplay BOOL,
+  html_node_lang STRING
+> LANGUAGE js AS '''
+var result = {};
+try {
+    var almanac = JSON.parse(almanac_string); 
```

Comment:
Nit: Inconsistent indentation. 4 spaces here but 2 below. Use 2 for consistency.

Reply:
Another code theft issue. I've noticed that Studio Code does deal with some indention improvements for JS, but it bases its rule on the current code. So sometimes it auto indents in different ways. I basically could not be bothered going through all those scripts trying to get them all to follow the same indentation pattern. Imagine if I had 100s of script to tidy up!